homeassistant/helpers/update_coordinator.py

```diff
@@ -148,21 +150,22 @@ def async_add_listener(
     ) -> Callable[[], None]:
         """Listen for data updates."""
         schedule_refresh = not self._listeners
-
-        @callback
-        def remove_listener() -> None:
-            """Remove update listener."""
-            self._listeners.pop(remove_listener)
-            if not self._listeners:
-                self._unschedule_refresh()
-
-        self._listeners[remove_listener] = (update_callback, context)
+        self._listener_id += 1
+        self._listeners[self._listener_id] = (update_callback, context)
```

Comment:
I used an incrementing ID here since `context` is typed to `Any` which means it doesn't have to hashable